File: pixleesdk/src/main/java/com/pixlee/pixleesdk/data/repository/BasicRepository.java
package com.pixlee.pixleesdk.data.repository;

import android.util.Base64;

import com.pixlee.pixleesdk.PXLClient;
import com.pixlee.pixleesdk.PXLPhoto;
import com.pixlee.pixleesdk.data.MediaResult;
import com.pixlee.pixleesdk.data.PhotoResult;
import com.pixlee.pixleesdk.data.api.BasicAPI;
import com.pixlee.pixleesdk.network.HMAC;
import com.pixlee.pixleesdk.network.multiparts.MultipartUtil;

import org.json.JSONObject;

import java.io.File;
import java.io.IOException;
import java.nio.file.Files;
import java.security.InvalidKeyException;
import java.security.NoSuchAlgorithmException;
import java.util.ArrayList;
import java.util.List;

import javax.crypto.Mac;
import javax.crypto.spec.SecretKeySpec;

import okhttp3.MediaType;
import okhttp3.MultipartBody;
import okhttp3.RequestBody;
import retrofit2.Call;

/**
 * This object loads data from and uploads data to the server using BasicAPI.java, a Retrofit HTTP API class.
 */
public class BasicRepository implements BasicDataSource {
    BasicAPI api;

    public BasicRepository(BasicAPI api) {
        this.api = api;
    }

    private String getSignature(JSONObject json) {
        String signature = null;
        try {
            signature = HMAC.computeHmac(json.toString().replace("\\/", "/"), PXLClient.secretKey);
        } catch (NoSuchAlgorithmException e) {
            e.printStackTrace();
        } catch (InvalidKeyException e) {
            e.printStackTrace();
        }
        return signature;
    }

    @Override
    public Call<PhotoResult> getPhotosWithSKU(String sku, String api_key, String filters, String sort, int per_page, int page) {
        return api.getPhotosWithSKU(sku, api_key, filters, sort, per_page, page);
    }

    @Override
    public Call<PhotoResult> getPhotosWithID(String album_id, String api_key, String filters, String sort, int per_page, int page) {
        return api.getPhotosWithID(album_id, api_key, filters, sort, per_page, page);
    }

    @Override
    public Call<PXLPhoto> getMedia(String album_photo_id, String api_key) {
        return api.getMedia(album_photo_id, api_key);
    }

    @Override
    public Call<MediaResult> postMedia(JSONObject json) {
        if (PXLClient.secretKey == null) {
            throw new IllegalArgumentException("no secretKey, please set secretKey before start");
        }

        RequestBody body = RequestBody.create(MediaType.parse("application/json; charset=utf-8"), json.toString());
        return api.postMedia(getSignature(json), PXLClient.apiKey, body);
    }

    @Override
    public Call<MediaResult> uploadImage(JSONObject json, String filePath) {
        List<MultipartBody.Part> bodyList = new ArrayList<>();
        File photo = new File(filePath);
        bodyList.add(new MultipartUtil().getMultipartBody("file", photo));
        bodyList.add(MultipartBody.Part.createFormData("json", json.toString()));
        return api.uploadImage(getSignature(json), PXLClient.apiKey, bodyList);
    }
}

File: pixleesdk/src/main/java/com/pixlee/pixleesdk/data/repository/BasicRepository.kt
package com.pixlee.pixleesdk.data.repository

import com.pixlee.pixleesdk.client.PXLClient
import com.pixlee.pixleesdk.data.PXLPhoto
import com.pixlee.pixleesdk.data.MediaResult
import com.pixlee.pixleesdk.data.PhotoResult
import com.pixlee.pixleesdk.data.api.BasicAPI
import com.pixlee.pixleesdk.network.multiparts.MultipartUtil
import com.pixlee.pixleesdk.util.toHMAC
import okhttp3.MultipartBody
import okhttp3.RequestBody.Companion.toRequestBody
import org.json.JSONObject
import retrofit2.Call
import java.io.File
import java.util.*

/**
 * Created by sungjun on 9/11/20.
 */
/**
 * The entry point for accessing basic data.
 */
interface BasicDataSource {
    /**
     * @param sku: product's sku
     * @param filter: filtering options
     * @param sort: sort options
     * @param per_page: how many PXLPhotos you want to receive in a response
     * @param page: this is for pagination, range [(]1 - N], example: first page is 1, the next page is 2, 3, 4, ...
     * @param region_id: region id to get data from a specific region
     * @return PhotoResult
     */
    fun getPhotosWithSKU(
            sku: String,
            filters: String?,
            sort: String?,
            per_page: Int,
            page: Int,
            region_id: Int?
    ): Call<PhotoResult>

    /**
     * @param album_id: album id
     * @param filter: filtering options
     * @param sort: sort options
     * @param per_page: how many PXLPhotos you want to receive in a response
     * @param page: this is for pagination, range [(]1 - N], example: first page is 1, the next page is 2, 3, 4, ...
     * @param region_id: region id to get data from a specific region
     * * @return PhotoResult
     */
    fun getPhotosWithID(
            album_id: String,
            filters: String?,
            sort: String?,
            per_page: Int,
            page: Int,
            region_id: Int?
    ): Call<PhotoResult>

    /**
     * @param album_photo_id: This is returns PXLPhoto with album_photo_id which can be discovered in PXLPhoto.albumPhotoId
     * @return PXLPhoto
     */
    fun getMedia(album_photo_id: String): Call<PXLPhoto>

    /**
     * @param album_photo_id: This is returns PXLPhoto with album_photo_id which can be discovered in PXLPhoto.albumPhotoId
     * @param region_id: region id to get data from a specific region
     * @return PXLPhoto
     */
    fun getPhoto(album_photo_id: String, region_id:Int?): Call<PXLPhoto>

    /**
     * Requests the next page of photos from the Pixlee album. Make sure to set perPage,
     * sort order, and filter options before calling.
     *
     * @param json: this should have these elements
     *              - title: title or caption of the photo being uploaded
     *              - email: email address of the submitting user
     *              - username: username of the submitting user
     *              - approved: boolean specifying whether the photo should be marked as approved on upload
     *              - photoURI: the URI of the photo/video being submitted (must be a public URI)
     */
    fun postMediaWithURI(json: JSONObject): Call<MediaResult>

    /**
     * Requests the next page of photos from the Pixlee album. Make sure to set perPage,
     * sort order, and filter options before calling.
     *
     * @param json: this has these values
     *              - title: title or caption of the photo being uploaded
     *              - email: email address of the submitting user
     *              - username: username of the submitting user
     *              - approved: boolean specifying whether the photo should be marked as approved on upload
     *              - localMediaPath: a local path of the media(photo or video)
     */
    fun postMediaWithURI(json: JSONObject, filePath: String): Call<MediaResult>
}

/**
 * This object loads data from and uploads data to the server using BasicAPI.java, a Retrofit HTTP API class.
 */
class BasicRepository(var api: BasicAPI) : BasicDataSource {
    override fun getPhotosWithSKU(sku: String, filters: String?, sort: String?, per_page: Int, page: Int, region_id: Int?): Call<PhotoResult> {
        return api.getPhotosWithSKU(sku, PXLClient.apiKey, filters, sort, per_page, page, region_id)
    }

    override fun getPhotosWithID(album_id: String, filters: String?, sort: String?, per_page: Int, page: Int, region_id: Int?): Call<PhotoResult> {
        return api.getPhotosWithID(album_id, PXLClient.apiKey, filters, sort, per_page, page, region_id)
    }

    override fun getMedia(album_photo_id: String): Call<PXLPhoto> {
        return api.getMedia(album_photo_id, PXLClient.apiKey)
    }

    override fun getPhoto(album_photo_id: String, region_id: Int?): Call<PXLPhoto> {
        return api.getPhoto(album_photo_id, PXLClient.apiKey, region_id)
    }

    override fun postMediaWithURI(json: JSONObject): Call<MediaResult> {
        return api.postMediaWithURI(json.toHMAC(), PXLClient.apiKey, json.toString().toRequestBody(PXLClient.mediaType))
    }

    override fun postMediaWithURI(json: JSONObject, filePath: String): Call<MediaResult> {
        val bodyList: MutableList<MultipartBody.Part> = ArrayList()
        val photo = File(filePath)
        bodyList.add(MultipartUtil().getMultipartBody("file", photo))
        bodyList.add(MultipartBody.Part.createFormData("json", json.toString()))
        return api.postMediaWithFile(json.toHMAC(), PXLClient.apiKey, bodyList)
    }
}
